atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/OracleShrinkExecutor.java

```diff
@@ -0,0 +1,144 @@
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Stopwatch;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleDdlConfig;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleTableNameGetter;
+import com.palantir.atlasdb.keyvalue.impl.TableMappingNotFoundException;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.exception.PalantirSqlException;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.nexus.db.sql.SqlConnection;
+
+public class OracleShrinkExecutor {
+    private static final Logger log = LoggerFactory.getLogger(OracleShrinkExecutor.class);
+
+    private final SqlConnectionSupplier connectionPool;
+    private final ExecutorService executorService;
+    private final OracleTableNameGetter oracleTableNameGetter;
+    private OracleDdlConfig oracleDdlConfig;
+
+    private static Future<Boolean> previousShrinkFuture;
+
+    public OracleShrinkExecutor(
+            SqlConnectionSupplier connectionPool,
+            ExecutorService executorService,
+            OracleTableNameGetter oracleTableNameGetter,
+            OracleDdlConfig oracleDdlConfig) {
+        this.connectionPool = connectionPool;
+        this.executorService = executorService;
+        this.oracleTableNameGetter = oracleTableNameGetter;
+        this.oracleDdlConfig = oracleDdlConfig;
+    }
+
+    public void shrinkAsync(TableReference tableRef) {
+        final long secondsToWaitBeforeShrink = getSecondsToWaitBeforeShrink();
+        previousShrinkFuture = executorService.submit(
+                () -> shrinkCompactFollowedByShrink(tableRef, secondsToWaitBeforeShrink));
+
+    }
+
+    private long getSecondsToWaitBeforeShrink() {
+        long secondsToWaitBeforeShrink = oracleDdlConfig.shrinkPauseSeconds();
+        if (previousShrinkFuture != null) {
+            try {
+                if (!previousShrinkFuture.get()) {
+                    secondsToWaitBeforeShrink =
+                            secondsToWaitBeforeShrink + oracleDdlConfig.shrinkPauseOnFailureSeconds();
+                }
+            } catch (InterruptedException | ExecutionException e) {
+                secondsToWaitBeforeShrink = secondsToWaitBeforeShrink + oracleDdlConfig.shrinkPauseOnFailureSeconds();
+            }
+        }
+        return secondsToWaitBeforeShrink;
+    }
+
+    private boolean shrinkCompactFollowedByShrink(TableReference tableRef, long secondsToWaitBeforeShrink) {
+
+        try {
+            TimeUnit.SECONDS.sleep(secondsToWaitBeforeShrink);
+        } catch (InterruptedException e) {
+            log.warn("Skipping Shrink for table: {}", LoggingArgs.tableRef("tableToShrink", tableRef));
+            return false;
+        }
+
+        Stopwatch timer = Stopwatch.createStarted();
+        ConnectionSupplier conns = new ConnectionSupplier(connectionPool);
+        try {
```

Comment:
This method is overlong (it clocks in at 45 lines). Extracting the whole of the `try` block would make the length a more reasonable 30-ish, and separate the responsibilities of (a) actually doing the shrink, and (b) handling error cases.

Reply:
The same block is also repeated twice from lines 94-99 and 101-107, with minimal variation. Consider further extracting a helper method. There are various options here; you could go as far as having the function be `alterTable(conns, tableRef, "SHRINK SPACE(| COMPACT)")`

A bonus of doing this: you only have to add `SafeArg` to one log message!